### ISSUE TITLE
"Restart" and/or "RestartForceExitStatus" should work with oneshot services, too

### ISSUE THREAD

**Issue by enkore** [systemd/systemd#2582]
While `Restart=on-success` for a `Type=oneshot` service is just a bad idea and should be excluded, other `Restart` modes should work.

This would allow systemd to at least try to automatically recover from errors (like temporary failure to acquire a resource, e.g. temporary connection failures and similar issues) in oneshot services, too. This would be mainly relevant to services run by timers.

I'm not familiar with the systemd codebase, so I can only guess how a patch would look like. The restriction itself is obviously implemented in core/service.c:482-490, whether simply removing / altering those lines (to exclude "Restart=on-success") would suffice I can not say.

**Comment by binarybana** [comment]
This would more closely match what most users expect from oneshot services, and prevents the [brittle  workarounds](https://unix.stackexchange.com/questions/272313/systemd-run-a-job-at-startup-successfully-to-completion/272650) needed to emulate this functionality otherwise.

**Comment by arvidjaar** [comment]
If service can detect condition when start attempt needs to be repeated, why it cannot simply repeat it itself?

**Comment by binarybana** [comment]
~~@arvidjaar Is that a technical question or a suggestion? If the latter, then that seems like an acceptable solution.~~

**Comment by enkore** [comment]
> If service can detect condition when start attempt needs to be repeated, why it cannot simply repeat it itself?

I.e. instead of the service manager managing the service the service shall manage itself?

**Comment by binarybana** [comment]
`oneshot` services managing their own restart policy (in a bash retry loop or similar) is similar to `simple` services managing their own restart policies: while it is entirely possible to do this, why not instead solve this _once_ at the service manager (systemd) layer and all downstream services easier to understand and less brittle?

**Comment by redbaron** [comment]
Would be nice to hear any update/opinion on this issue from systemd devs? @poettering ?

**Comment by neingeist** [comment]
I'd also like to see "Restart" for oneshot services. On my Fedora systems, the dnf-makecache.service (oneshot) sporadically fails and it would be great if I could just configure it to restart on fail.

**Comment by cognifloyd** [comment]
I have a set of Type=oneshot services that BindsTo=P.service. P in turn BindsTo=<a usb device> (with the udev rules to trigger systemd activation of P.service). Sadly, if the device is unplugged at just the right time, or something else is blocking iptables or several other race conditions, one or more of the oneshot services will fail. I want them to try again for as long as the device is plugged in. As is now, if they enter a failed state, and I unplug/plug the device back in, none of the services start again. Unplugging the device should "reset" any of the failed states. My workaround for that was to set Restart=on-failure on the Type=oneshot services, however systemd doesn't support that. Please allow Restart= for Type=oneshot services.
_Edit: If helpful, [these](https://bitbucket.org/theatro-noc/gadget-pppd-units) are the units I'm dealing with_

**Comment by arvidjaar** [comment]
> As is now, if they enter a failed state, and I unplug/plug the device back in, none of the services start again. 

I do not see how `Restart=` would help with it. `Restart=` would make it be restarted again and again even though device is unplugged. Which unlikely is what you intend. Note that even if service is in Failed state systemd will attempt to start it again when requested (consider it sort of "auto reset"). See example below. So as long as plugging device again triggers some top level service that pulls in your `Type=oneshot` service I expect it to work.

Note that `Requires=test.service` below is only to make `top.service` fail. Otherwise it remains started from systemd PoV. I suspect something like this happens in your case (service that should start `Type=oneshot` service remains running).

```
bor@bor-Latitude-E5450:~/src/systemd$ cat ~/.local/share/systemd/user/top.service
[Unit]
Description=test dependency
Requires=test.service
After=test.service

[Service]
Type=oneshot
RemainAfterExit=true
ExecStart=/bin/true
bor@bor-Latitude-E5450:~/src/systemd$ cat ~/.local/share/systemd/user/test.service
[Unit]
Decription=service that fails

[Service]
Type=oneshot
RemainAfterExit=true
ExecStart=/bin/sh -c 'if test -e /tmp/test-started; then /bin/true; else touch /tmp/test-started; /bin/false; fi'
ExecStop=/bin/sh -c 'if test -e /tmp/test-started; then rm -f /tmp/test-started; /bin/true; else /bin/false; fi'
bor@bor-Latitude-E5450:~/src/systemd$ systemctl --user start top.service
A dependency job for top.service failed. See 'journalctl -xe' for details.
bor@bor-Latitude-E5450:~/src/systemd$ systemctl --user status top.service
● top.service - test dependency
   Loaded: loaded (/home/bor/.local/share/systemd/user/top.service; static; vendor preset: enab
   Active: inactive (dead)

мар 05 10:56:51 bor-Latitude-E5450 systemd[1652]: Dependency failed for test dependency.
мар 05 10:56:51 bor-Latitude-E5450 systemd[1652]: top.service: Job top.service/start failed wit
bor@bor-Latitude-E5450:~/src/systemd$ systemctl --user status test.service
● test.service
   Loaded: loaded (/home/bor/.local/share/systemd/user/test.service; static; vendor preset: ena
   Active: failed (Result: exit-code) since Вс 2017-03-05 10:56:51 MSK; 14s ago
  Process: 31410 ExecStart=/bin/sh -c if test -e /tmp/test-started; then /bin/true; else touch 
 Main PID: 31410 (code=exited, status=1/FAILURE)

мар 05 10:56:51 bor-Latitude-E5450 systemd[1652]: Starting test.service...
мар 05 10:56:51 bor-Latitude-E5450 systemd[1652]: test.service: Main process exited, code=exite
мар 05 10:56:51 bor-Latitude-E5450 systemd[1652]: Failed to start test.service.
мар 05 10:56:51 bor-Latitude-E5450 systemd[1652]: test.service: Unit entered failed state.
мар 05 10:56:51 bor-Latitude-E5450 systemd[1652]: test.service: Failed with result 'exit-code'.
bor@bor-Latitude-E5450:~/src/systemd$ systemctl --user start top.service
bor@bor-Latitude-E5450:~/src/systemd$ systemctl --user status top.service
● top.service - test dependency
   Loaded: loaded (/home/bor/.local/share/systemd/user/top.service; static; vendor preset: enab
   Active: active (exited) since Вс 2017-03-05 10:57:08 MSK; 2s ago
  Process: 31429 ExecStart=/bin/true (code=exited, status=0/SUCCESS)
 Main PID: 31429 (code=exited, status=0/SUCCESS)

мар 05 10:57:08 bor-Latitude-E5450 systemd[1652]: Starting test dependency...
мар 05 10:57:08 bor-Latitude-E5450 systemd[1652]: Started test dependency.
bor@bor-Latitude-E5450:~/src/systemd$ systemctl --user status test.service
● test.service
   Loaded: loaded (/home/bor/.local/share/systemd/user/test.service; static; vendor preset: ena
   Active: active (exited) since Вс 2017-03-05 10:57:08 MSK; 5s ago
  Process: 31427 ExecStart=/bin/sh -c if test -e /tmp/test-started; then /bin/true; else touch 
 Main PID: 31427 (code=exited, status=0/SUCCESS)
   CGroup: /user.slice/user-1001.slice/user@1001.service/test.service

мар 05 10:57:08 bor-Latitude-E5450 systemd[1652]: Starting test.service...
мар 05 10:57:08 bor-Latitude-E5450 systemd[1652]: Started test.service.
```

**Comment by arvidjaar** [comment]
> Otherwise it remains started from systemd PoV.

Which still causes systemd to initiate start job for dependent services (cf. https://github.com/systemd/systemd/issues/4221).

**Comment by llamahunter** [comment]
oneshot services should support restart.  I have a oneshot that relies on a sometimes flaky remote service.  Very rarely, that remote service doesn't respond correctly (I suspect it is overloaded), but a subsequent try or two works.  I would like the oneshot to keep retrying until it succeeds. If this particular oneshot does not execute, many things in the system won't work, so it blocks other units from starting until its successful.

Saying 'fix the external service' is not really a doable solution.  I don't have much control over it.

**Comment by businessBoris** [comment]
I'm curious as to the reason for _not_ including it.
I have a simple oneshot service which sometimes fails because it cannot always resolve a particularly hostname.

**Comment by rickalm** [comment]
Adding to the conversation. I have a oneshot service which is designed to discover and format a disk on initial boot.

besides the Requires= and After= in the [unit] I have

ConditionPathExists=/dev/disk/by-label/DISK1
ConditionPathExists=!/dev/disk/by-label/DISK2

In this case this oneshot is intended to format a disk and name it DISK2 but only after the other oneshot job has succeeded in formatting DISK1. Since it takes linux a few seconds to update the by-label tree there is a period in time which the first job has completed, but the labels have not been updated and proceeding is not advised.

I intended to add the following assuming it would cause systemd to re-evaluate the conditionals until it qualified to be run. 

[Service]
RestartSec=30
Restart=on-failure

Like the others above I could add a "while ! -e /dev/disk/by-label/DISK1; do sleep 1; done" to my start command but its counter-intuitive to add dependency management to the command rather than let systemd manage that

**Comment by alexrudd** [comment]
I have the same issue as @llamahunter

I can almost get the behavior I need by using `Type=simple` `restart=on-failure` `RemainAfterExit=yes` but then all depending services will try to start (and then fail) before the initial service has a chance to fully complete.

I assume this is because you only get the following behavior with oneshot: "it is expected that the process has to exit before systemd starts follow-up units"

**Comment by virtualdxs** [comment]
Any update on this? Is there a reason it's not implemented?

**Comment by ipatrol** [comment]
I would like this too. Is there a particular technical reason it's not implemented, or is it just a time/labor issue? Sadly I cannot help with that, because my C-fu is not up to the task.

**Comment by poettering** [comment]
> Any update on this? Is there a reason it's not implemented?

as it is always with feature requests: if you want something, then send a PR, or convince somebody else to do it for you. Conceptually there's no opposition to making this work. it just needs somebody to hack this up, and maybe add a test for this, and then submit it as PR, and stick around long enough to shepherd it through the review process.

**Comment by mlushpenko** [comment]
+ 1 for this, I see there is a merge request but looks like it got stuck a few months ago. Is there no some core team who can finish this as it looks like quite a few people want this? Probably I will have some other solution by the time it is implemented, but just curious

**Comment by robermorales** [comment]
We need to talk before, if we wanna to open the option to `on-failure` or for all the rest of Restart types. https://github.com/systemd/systemd/pull/7474#issuecomment-349330650

Since I am not whole idea of the basecode and I am not aware of the potential failures derived from such a change, I preferred to stop working in the PR.

I still need the change.

**Comment by corsac-s** [comment]
Related but not identical to this issue: ExecStopPost= doesn't seem to work with Type=oneshot, so it's not possible right now to nicely handle failure of an ExecStart= job.
I'm unsure of the side effects, but it would be nice to have. I can open a separate issue but it looked similar enough.

**Comment by robermorales** [comment]
I think the topic is similar enough, @corsac-s 

So I think, that it is better changing the issue title to express a more generic intention:

"Review enabling combinations of options and types that are disabled now, but can be useful" e.g.

Then, we can start talking about a matrix with options and types, and let the pairs with unambiguous semantic to be enabled, even if the use case can seem weird.

**Comment by jessebl** [comment]
Has there been any internal word on getting this implemented?

**Comment by SjonHortensius** [comment]
@jessebl If you want this done properly you'll have to build it yourself. As a work-around, I'm using:

`/etc/systemd/system/fail-notify@.service`
```bash
[Unit]
Description=Restart a failed service, implements Restart on-failure for simple/oneshot
# Enable by specifying OnFailure=fail-notify@%n.service AND #RestartSec=xxx

[Service]
Type=oneshot
ExecStartPre=/bin/sh -c 'systemctl set-environment SEC="`systemctl cat %I|grep -Po \\"^#RestartSec=\\K(\\d)*(?=s?$$)\\"`"'
   ExecStart=/bin/sh -c 'sleep ${SEC:-0.1}; systemctl restart %I'
```

`/etc/systemd/system/service.conf.d/restart.conf`
```bash
[Unit]
OnFailure=restart-on-failure@%n
#RestartSec=3
```

**Comment by robermorales** [comment]
I think that managing the restart policy is (should be) one of the core intentions of a service manager like systemd. Solving the restart policy at the service level is far from perfect.

**Comment by diasjorge** [comment]
Any possibility to move forward the issue? It's been stalled almost a year please.

**Comment by edp-d6labs** [comment]
What is the status of this issue? I also need restart capability with oneshot services.

**Comment by diasjorge** [comment]
I've implemented some retry logic in my startup scripts and dealt with it. There was a PR https://github.com/systemd/systemd/pull/7474#issuecomment-349330650 but it was closed as it would need to be implemented for all services. I don't think the maintainers are interested or probably have time for it. If someone implemented a better PR would it have a chance?

**Comment by anitazha** [comment]
PR #13754 was just merged that will cover `Restart=` for oneshot. Since there was an "or" in the issue title and it seems like people primarily just wanted `Restart=`, I'm closing this out.

**Comment by corsac-s** [comment]
> PR #13754 was just merged that will cover `Restart=` for oneshot. Since there was an "or" in the issue title and it seems like people primarily just wanted `Restart=`, I'm closing this out.

Unfortunately I don't think that cover my use case:

> Related but not identical to this issue: ExecStopPost= doesn't seem to work with Type=oneshot, so it's not possible right now to nicely handle failure of an ExecStart= job.
> I'm unsure of the side effects, but it would be nice to have. I can open a separate issue but it looked similar enough.

Should I open a new issue then?

**Comment by anitazha** [comment]
@corsac-s Sorry I missed that since the thread was so lengthy. Please open a new issue for that request.

**Comment by corsac-s** [comment]
> @corsac-s Sorry I missed that since the thread was so lengthy. Please open a new issue for that request.

Done at #13796
